File: freedom.go
// Package freedom 是一个从零自研的 WebView 桌面壳子框架（对标 Wails / Tauri）。
//
// Freedom 2.0 架构：
//   - shell：跨平台渲染壳。复用各系统自带 WebView 内核（Windows WebView2 / macOS
//     WKWebView / Linux WebKitGTK），通过跨平台库 webview_go 绑定，自身不携带浏览器
//     内核。壳负责窗口生命周期、前端资源 embed、后端进程管理、IPC 路由。一份壳代码
//     编译三平台。
//   - backend：后端抽象。任意时刻绑定一个后端，前端 window.freedom.call 全部路由过去。
//     内置两种实现：
//       * EmbedBackend：Go 方法直接注册在壳进程内（app.Bind，兼容 v1 用法）。
//       * ProcBackend：后端是任意语言实现的独立进程（Go/Rust/Python/Node/Java…），
//         通过 stdin/stdout 上的 NDJSON/JSON-RPC 与壳通信。协议语言无关，
//         换一种后端语言无需改壳与前端。
//   - ipc：双向桥接。前端 window.freedom.call(method, ...args) -> Promise；
//     后端 app.Emit(event, data) 向所有前端监听器推送事件。
//   - assets：前端编译产物（Vite 单文件 HTML）通过 go:embed 嵌入二进制，运行时
//     直接 SetHtml 从内存加载，无需本地 HTTP 端口。
package freedom

import (
	"encoding/json"
	"fmt"
	"os"
	"sync"
	"unsafe"

	webview "github.com/webview/webview_go"
)

// TitleBarMode 描述窗口标题栏策略。
type TitleBarMode string

const (
	// TitleBarNative 保留系统原生标题栏（非默认；默认见 TitleBarFrameless）。
	TitleBarNative TitleBarMode = "native"
	// TitleBarHidden 隐藏标题栏视觉，但保留系统原生最小化 / 最大化 / 关闭按钮。
	// 当前仅 Windows 生效（基于 DWM 扩展实现）；macOS / Linux 上回退为原生标题栏。
	TitleBarHidden TitleBarMode = "hidden"
	// TitleBarFrameless 完全无边框，客户区铺满整个窗口。
	// 最小化 / 最大化 / 关闭按钮需由前端自绘（通过 window.freedom.window.* 控制），
	// 三平台行为一致。
	TitleBarFrameless TitleBarMode = "frameless"
)

// Config 描述一个 Freedom 应用的窗口与运行配置。
type Config struct {
	// Title 是窗口标题。
	Title string
	// TitleBar 指定标题栏策略（默认 TitleBarFrameless）。可通过 freedom CLI 一键切换。
	TitleBar TitleBarMode
	// Width / Height 是窗口初始尺寸（像素）。
	Width  int
	Height int
	// Center 为 true 时窗口在屏幕居中（Windows 生效；macOS/Linux 由窗口管理器决定）。
	Center bool
	// MinWidth / MinHeight 为窗口最小尺寸。两个字段均 >0 时作为一组约束生效
	//（webview 的 HintMin 按整组应用）；任一 <=0 表示不启用最小尺寸限制。
	MinWidth  int
	MinHeight int
	// Debug 为 true 时开启 WebView 开发者工具（目标平台支持时）。
	Debug bool
	// Backend 指定后端适配器。为空时默认使用内嵌 Go 后端（配合 Bind 使用）。
	Backend Backend
	// HTML 返回要加载到窗口的前端页面内容（内存加载，无本地端口）。
	// 为 nil 时使用框架内置的默认占位页。
	HTML func() (string, error)
}

// App 是 Freedom 应用实例。
type App struct {
	cfg     Config
	view    webview.WebView
	viewMu  sync.RWMutex // 保护 view：Emit/Quit/WindowHandle 承诺可从任意 goroutine 调用
	backend Backend
	onReady func(a *App)
}

// setView / getView 提供 view 字段的并发安全访问。
func (a *App) setView(w webview.WebView) {
	a.viewMu.Lock()
	a.view = w
	a.viewMu.Unlock()
}

func (a *App) getView() webview.WebView {
	a.viewMu.RLock()
	defer a.viewMu.RUnlock()
	return a.view
}

// New 创建并初始化一个 Freedom 应用。调用 Run() 之前不会显示窗口。
func New(cfg Config) *App {
	if cfg.Title == "" {
		cfg.Title = "Freedom App"
	}
	// M1：默认无边框（frameless），与 freedom-cli 模板默认一致。
	// 此前默认 native 而 CLI 默认 frameless，resources 缺失/配置不完整时
	// 会出现"系统原生标题栏 + 页面自绘标题栏"双标题栏错位。
	if cfg.TitleBar == "" {
		cfg.TitleBar = TitleBarFrameless
	}
	if cfg.Width <= 0 {
		cfg.Width = 1024
	}
	if cfg.Height <= 0 {
		cfg.Height = 768
	}
	if cfg.Backend == nil {
		cfg.Backend = NewEmbedBackend()
	}
	return &App{cfg: cfg, backend: cfg.Backend}
}

// OnReady 注册一个回调，在窗口与桥接层就绪、页面加载前执行。
// 回调内可以安全调用 Emit 向已注入的页面发送初始化事件。
func (a *App) OnReady(fn func(a *App)) {
	a.onReady = fn
}

// Bind 把后端 Go 方法暴露给前端。仅当后端为内嵌 Go 后端（默认）时有效；
// 进程后端的方法由后端进程自身注册，此处调用会返回错误。
//
//   - fn 必须是函数。
//   - 参数与返回值通过 JSON 编解码，前端用 window.freedom.call(name, ...args) 调用，
//     返回 Promise（resolve 为返回值，reject 为 error 的字符串表示）。
//   - 返回值约定：可返回 (T, error) 或仅 (T) 或仅 error 或空。
func (a *App) Bind(name string, fn interface{}) error {
	eb, ok := a.backend.(*EmbedBackend)
	if !ok {
		return fmt.Errorf("freedom: Bind 仅适用于内嵌 Go 后端；当前后端为 %T，方法请在进程后端中注册", a.backend)
	}
	return eb.Bind(name, fn)
}

// Unbind 移除先前 Bind 的方法（内嵌后端）。
func (a *App) Unbind(name string) {
	if eb, ok := a.backend.(*EmbedBackend); ok {
		eb.Unbind(name)
	}
}

// Run 启动窗口并进入主事件循环，阻塞直到窗口被关闭。
func (a *App) Run() {
	html, err := a.resolveHTML()
	if err != nil {
		fmt.Printf("freedom: failed to resolve HTML: %v\n", err)
		return
	}

	w := webview.New(a.cfg.Debug)
	if w == nil {
		fmt.Println("freedom: failed to create webview")
		return
	}
	a.setView(w)
	defer func() {
		if err := a.backend.Close(); err != nil {
			fmt.Fprintf(os.Stderr, "freedom: backend close: %v\n", err)
		}
		// 销毁前摘除引用：Run 返回后 Emit/Quit/WindowHandle 不得再触碰已销毁的 webview。
		a.setView(nil)
		w.Destroy()
	}()

	// 进程后端：启动后端进程并把其推送的事件转发到前端。
	if pb, ok := a.backend.(*ProcBackend); ok {
		pb.OnEvent(func(event string, data interface{}) {
			a.Emit(event, data)
		})
		if err := pb.start(); err != nil {
			fmt.Printf("freedom: backend start failed: %v\n", err)
			return
		}
	}

	w.SetTitle(a.cfg.Title)
	w.SetSize(a.cfg.Width, a.cfg.Height, webview.HintNone)
	// 最小尺寸：任一项 >0 即生效；未设置的一项传 0 表示该维度不限制。
	if a.cfg.MinWidth > 0 || a.cfg.MinHeight > 0 {
		w.SetSize(a.cfg.MinWidth, a.cfg.MinHeight, webview.HintMin)
	}
	a.applyCenter()
	a.applyTitleBar()

	// 注入前端 SDK：window.freedom 全局对象。
	w.Init(jsSDK)

	// 绑定 IPC 桥接入口：前端 window.__freedom_bridge(method, paramsJson)。
	// webview_go 的 Bind 会让前端调用返回 Promise，Go 侧结果自动 JSON 序列化回传。
	if err := w.Bind("__freedom_bridge", a.bridge); err != nil {
		fmt.Printf("freedom: failed to bind bridge: %v\n", err)
		return
	}
	// 框架内置方法：健康检查。
	if err := w.Bind("__freedom__ping", func() string { return "pong" }); err != nil {
		fmt.Printf("freedom: failed to bind ping: %v\n", err)
		return
	}
	// 框架内置方法：窗口控制（无边框模式下前端自绘按钮使用）。
	// 支持动作：minimize / maximize / unmaximize / toggleMaximize / close / isMaximized / isFrameless。
	if err := w.Bind("__freedom_window", a.windowControl); err != nil {
		fmt.Printf("freedom: failed to bind window control: %v\n", err)
		return
	}
	// 框架内置方法：原生系统能力（wails v3 对标层）。
	// 支持方法：taskbar.*（进度/状态/角标）、window.backdrop/corner/borderColor、
	// dialog.message/open/save。平台实现分文件：syscap_windows.go / syscap_other.go。
	if err := w.Bind("__freedom_sys", a.sysCapCall); err != nil {
		fmt.Printf("freedom: failed to bind sys capability: %v\n", err)
		return
	}
	// 框架内置方法：系统托盘与原生菜单栏。
	// 支持方法：tray.create/destroy/tooltip/menu、menu.set。
	// 平台实现分文件：tray_windows.go / tray_other.go。
	if err := w.Bind("__freedom_tray", a.trayCall); err != nil {
		fmt.Printf("freedom: failed to bind tray: %v\n", err)
		return
	}
	// 框架内置方法：页面就绪回调（H3）。前端 SDK 在 DOMContentLoaded 后调用，
	// 保证 onReady 触发时页面已加载、监听器已注册；此前 onReady 在 SetHtml 前
	// 触发，期间 Emit 的初始化事件因 window.freedom 尚未建立而丢失。
	// 页面可能因导航/重载多次触发 DOMContentLoaded，用 once 兜底只执行一次。
	var readyOnce sync.Once
	if err := w.Bind("__freedom__ready", func() {
		readyOnce.Do(func() {
			if a.onReady != nil {
				a.onReady(a)
			}
		})
	}); err != nil {
		fmt.Printf("freedom: failed to bind ready: %v\n", err)
		return
	}

	// 注：onReady 不再在 SetHtml 前触发（H3），改由前端 SDK 在页面
	// DOMContentLoaded 后经 __freedom__ready 回调，保证初始化事件不丢失。
	w.SetHtml(html)
	w.Run()
}

// bridge 是前端调用后端的统一入口（JSON-RPC 风格）。
// 前端 SDK 通过 window.__freedom_bridge(method, paramsJson) 调用，
// 桥接层把请求转发给当前绑定的后端（内嵌 Go 方法或任意语言进程）。
func (a *App) bridge(method string, paramsJSON string) (result json.RawMessage, err error) {
	// 防御后端方法 panic 导致整个窗口崩溃：统一转为错误回传前端。
	defer func() {
		if r := recover(); r != nil {
			result = nil
			err = fmt.Errorf("freedom: method %q panicked: %v", method, r)
		}
	}()

	// 框架内置健康检查：__freedom__ping 同时以 JS 直调全局函数与桥接路由两种方式
	// 暴露（占位页/文档示例走桥接路由），此处兜底保证两条链路行为一致。
	if method == "__freedom__ping" {
		return json.RawMessage(`"pong"`), nil
	}

	var params []json.RawMessage
	if len(paramsJSON) > 0 && paramsJSON != "null" {
		if err := json.Unmarshal([]byte(paramsJSON), &params); err != nil {
			return nil, fmt.Errorf("freedom: method %q: invalid params: %w", method, err)
		}
	}

	raw, err := a.backend.Handle(method, params)
	if err != nil {
		return nil, err
	}
	data, err := json.Marshal(raw)
	if err != nil {
		return nil, fmt.Errorf("freedom: method %q: cannot marshal result: %w", method, err)
	}
	return json.RawMessage(data), nil
}

// Emit 把事件推送到前端。前端通过 window.freedom.on(event, cb) 订阅。
// 线程安全：可从任意 goroutine 调用（进程后端推送的事件亦经由本函数）。
func (a *App) Emit(event string, data interface{}) {
	view := a.getView()
	if view == nil {
		return
	}
	eb, _ := json.Marshal(event)
	db, err := json.Marshal(data)
	if err != nil {
		db = []byte("null")
	}
	js := "window.freedom && window.freedom.emit(" + string(eb) + "," + string(db) + ");"
	view.Dispatch(func() {
		view.Eval(js)
	})
}

// Quit 关闭窗口并退出应用。可从任意 goroutine 调用。
func (a *App) Quit() {
	view := a.getView()
	if view != nil {
		view.Dispatch(func() {
			view.Terminate()
		})
	}
}

// WindowHandle 返回底层原生窗口句柄（Windows 上为 HWND）。
func (a *App) WindowHandle() uintptr {
	view := a.getView()
	if view == nil {
		return 0
	}
	return uintptr(unsafe.Pointer(view.Window()))
}

// windowControl 处理前端 window.freedom.window.* 的窗口控制请求。
// 具体实现按平台分文件：window_windows.go（Windows）/ window_other.go（macOS、Linux）。
func (a *App) windowControl(action string) (result interface{}, err error) {
	// H1：windowControl 直接经 w.Bind 暴露给前端，不经 bridge 的 recover 兜底；
	// 平台实现（如 appIcon → dibToPNG 解析越界）一旦 panic 会直接崩掉整个壳进程。
	// 这里统一加 recover，把 panic 转成错误回传前端（页面 catch 后提示，进程不崩）。
	defer func() {
		if r := recover(); r != nil {
			result = nil
			err = fmt.Errorf("freedom: window action %q panicked: %v", action, r)
		}
	}()
	return windowControl(a.WindowHandle(), action, a.cfg.TitleBar)
}

// resolveHTML 依据配置返回页面内容。
func (a *App) resolveHTML() (string, error) {
	if a.cfg.HTML != nil {
		return a.cfg.HTML()
	}
	return defaultHTML, nil
}

File: assets/freedom.js
// Freedom 前端 SDK。
// window.__freedom_bridge 由 go-webview2 的 Bind 机制注册（异步桥接函数，返回 Promise），
// 注入脚本与 SDK 的执行先后顺序不固定，因此 SDK 采用"每次调用时动态读取"策略，
// 保证与注入顺序无关。
(function () {
  'use strict';

  function getBridge() {
    return (typeof window.__freedom_bridge === 'function') ? window.__freedom_bridge : null;
  }

  var listeners = {};

  // M1 异步桥接协议：__freedom_bridge(id, method, paramsJson) 只做投递（ack），
  // Go 侧完成后经 freedom.__resolve(id, {ok,result|error}) 回写本表登记的 Promise。
  var callSeq = 0;
  var pendingCalls = {};

  function call(method) {
    var params = Array.prototype.slice.call(arguments, 1);
    var bridge = getBridge();
    if (!bridge) {
      return Promise.reject(new Error('[freedom] 未检测到原生桥接（__freedom_bridge），当前不在桌面壳内运行。'));
    }
    var id = ++callSeq;
    return new Promise(function (resolve, reject) {
      pendingCalls[id] = { resolve: resolve, reject: reject };
      Promise.resolve(bridge(id, method, JSON.stringify(params))).catch(function (e) {
        if (pendingCalls[id]) {
          delete pendingCalls[id];
          reject(e);
        }
      });
    });
  }

  function on(event, cb) {
    if (typeof cb !== 'function') return function () {};
    (listeners[event] = listeners[event] || []).push(cb);
    return function () { off(event, cb); };
  }

  function off(event, cb) {
    var l = listeners[event] || [];
    var i = l.indexOf(cb);
    if (i >= 0) l.splice(i, 1);
  }

  // 一次性订阅：触发一次后自动退订（对标 Tauri listen+unlisten）。
  // 注意：必须用"注册进数组的那个 wrapper"退订——on() 返回的是 unlisten
  // 闭包而非回调本身，拿它去 off() 删不掉监听器（旧 bug，tests/sdk-surface 锁定）。
  function once(event, cb) {
    if (typeof cb !== 'function') return function () {};
    var fn = function (data) {
      off(event, fn);
      cb(data);
    };
    return on(event, fn);
  }

  // 后端通过 app.Emit(event, data) 触发（Eval 调用本函数）。
  function emit(event, data) {
    var l = listeners[event] || [];
    for (var i = 0; i < l.length; i++) {
      try { l[i](data); } catch (e) { /* 监听器异常不影响其余监听器 */ }
    }
  }

  // Go 侧异步回写入口（dispatch.go 的 resolveJS 经 Eval 调用）。
  // 拒绝值保持旧契约：webview_go 时代 catch 到的是错误消息字符串。
  function resolveCall(id, env) {
    var p = pendingCalls[id];
    if (!p) return; // 页面已重载/回写迟到：静默丢弃
    delete pendingCalls[id];
    if (env && env.ok) p.resolve(env.result);
    else p.reject((env && env.error) || '[freedom] call failed');
  }

  var freedom = {
    call: call,
    invoke: call,
    on: on,
    off: off,
    once: once,
    emit: emit,
    __resolve: resolveCall,
    window: {
      // 窗口控制（无边框/隐藏标题栏模式下前端自绘按钮使用）。
      // 全部经桥接调用原生实现，返回 Promise。
      minimize: function () { return windowAction('minimize'); },
      maximize: function () { return windowAction('maximize'); },
      unmaximize: function () { return windowAction('unmaximize'); },
      toggleMaximize: function () { return windowAction('toggleMaximize'); },
      // close(force)：force=true 时绕过 interceptClose 拦截直接关闭。
      close: function (force) { return windowAction('close', { force: !!force }); },
      isMaximized: function () { return windowAction('isMaximized'); },
      isFrameless: function () { return windowAction('isFrameless'); },
      // —— W1 对标 Tauri：位置 / 尺寸（物理像素）——
      setPosition: function (x, y) { return windowAction('setPosition', { x: x, y: y }); },
      setSize: function (width, height) { return windowAction('setSize', { width: width, height: height }); },
      getPosition: function () { return windowAction('getPosition'); },
      getSize: function () { return windowAction('getSize'); },
      innerSize: function () { return windowAction('innerSize'); },
      center: function () { return windowAction('center'); },
      setTitle: function (title) { return windowAction('setTitle', { title: String(title) }); },
      // —— 可见性 / 层级 / 焦点 ——
      show: function () { return windowAction('show'); },
      hide: function () { return windowAction('hide'); },
      focus: function () { return windowAction('focus'); },
      isVisible: function () { return windowAction('isVisible'); },
      isFocused: function () { return windowAction('isFocused'); },
      isMinimized: function () { return windowAction('isMinimized'); },
      setAlwaysOnTop: function (on) { return windowAction('setAlwaysOnTop', { on: !!on }); },
      setSkipTaskbar: function (on) { return windowAction('setSkipTaskbar', { on: !!on }); },
      // —— 行为开关 ——
      setResizable: function (on) { return windowAction('setResizable', { on: !!on }); },
      setMaximizable: function (on) { return windowAction('setMaximizable', { on: !!on }); },
      setMinimizable: function (on) { return windowAction('setMinimizable', { on: !!on }); },
      // —— 全屏 / 关闭拦截 ——
      setFullscreen: function (on) { return windowAction('setFullscreen', { on: !!on }); },
      enterFullscreen: function () { return windowAction('setFullscreen', { on: true }); },
      exitFullscreen: function () { return windowAction('setFullscreen', { on: false }); },
      isFullscreen: function () { return windowAction('isFullscreen'); },
      // 拦截系统/按钮关闭：开启后关闭动作变为 window.closeRequested 事件，
      // 前端处理后以 close(true) 放行或不做操作保持窗口存活。
      interceptClose: function (on) { return windowAction('interceptClose', { on: !!on }); },
      // 聚合信息：title/outerX/outerY/outerWidth/outerHeight/innerWidth/innerHeight/
      // scaleFactor/visible/focused/maximized/minimized/alwaysOnTop/skipTaskbar/frameless/fullscreen。
      getInfo: function () { return windowAction('getInfo'); },
      // 显示器列表（经 sys 桥接）：[{x,y,width,height,workX,workY,workWidth,workHeight,scaleFactor,isPrimary}]
      monitors: function () { return sysCall('window.monitors', {}); },
      // 窗口视觉效果（Win11）：backdrop: auto|none|solid|mica|acrylic；corner: round|roundSmall|square|default；borderColor: 0x00BBGGRR
      setBackdrop: function (mode) { return sysCall('window.backdrop', { mode: mode }); },
      setCorner: function (mode) { return sysCall('window.corner', { mode: mode }); },
      setBorderColor: function (color) { return sysCall('window.borderColor', { color: color }); },
      // —— M2 多窗口 ——
      // id()：本窗口标识（'main' 或 'w1'…）；create({title,width,height,center,url,html})：
      // 异步拉起次级窗口返回 {id}；list() 全部存活窗口；closeWindow/focusWindow(id)。
      // create 的页面来源三选一：url（外链）> html（内联字符串）> 复制主页面——
      // 主页面含开窗脚本时务必给 url 或 html，避免级联。
      // 注意：次级窗口无 sys/tray 桥，托盘/热键/状态记忆等平台单例只挂主窗口；
      // 次级窗口里 window.close() 关闭的是它自己。
      id: function () { return windowAction('id'); },
      list: function () { return windowAction('list'); },
      create: function (opts) { return windowAction('create', opts || {}); },
      closeWindow: function (id) { return windowAction('closeWindow', { id: id }); },
      focusWindow: function (id) { return windowAction('focusWindow', { id: id }); },
    },
    // ---- W2 系统集成命名空间（均经 __freedom_sys / __freedom_tray 桥接）----
    sys: sysCall,
    clipboard: {
      readText: function () { return sysCall('clipboard.read', {}); },
      writeText: function (text) { return sysCall('clipboard.write', { text: String(text) }); },
    },
    shell: {
      // 用系统默认程序打开 URL/文件（openExternal）。
      open: function (target) { return sysCall('shell.open', { target: String(target) }); },
    },
    notification: {
      show: function (title, body) { return sysCall('notification.show', { title: String(title), body: String(body || '') }); },
    },
    shortcut: {
      // combo 形如 "ctrl+alt+k"；触发时收到事件 freedom.on('shortcut.triggered', ({id}) => ...)。
      register: function (id, combo) { return sysCall('shortcut.register', { id: id, combo: combo }); },
      unregister: function (id) { return sysCall('shortcut.unregister', { id: id }); },
      list: function () { return sysCall('shortcut.list', {}); },
    },
    autostart: {
      isEnabled: function (name) { return sysCall('autostart.get', { name: name || '' }); },
      set: function (enabled, opts) { return sysCall('autostart.set', { name: (opts && opts.name) || '', enabled: !!enabled, args: (opts && opts.args) || '' }); },
      enable: function (opts) { return this.set(true, opts); },
      disable: function (opts) { return this.set(false, opts); },
    },
    protocol: {
      // URL Scheme 注册（deep link）；拉起参数经 'app.secondInstance' 事件或 app.launchArgs() 获取。
      register: function (scheme, displayName) { return sysCall('protocol.register', { scheme: scheme, name: displayName || '' }); },
      unregister: function (scheme) { return sysCall('protocol.unregister', { scheme: scheme }); },
    },
    app: {
      launchArgs: function () { return sysCall('app.launchArgs', {}); },
    },
    // ---- W3 数据层（对标 Tauri path / store / os / process）----
    // kind ∈ config|data|cache|temp|home|exe；name 省略时用 AppID（目录名已消毒）。
    path: function (kind, name) { return sysCall('path.get', { kind: kind, name: name || '' }); },
    store: {
      // 命名 JSON KV，落盘 <dataDir>/<name>.store.json，set 即持久化。
      load: function (name) { return sysCall('store.load', { store: name || '' }); },
      get: function (key, name) { return sysCall('store.get', { store: name || '', key: key }); },
      set: function (key, value, name) { return sysCall('store.set', { store: name || '', key: key, value: value }); },
      delete: function (key, name) { return sysCall('store.delete', { store: name || '', key: key }); },
      keys: function (name) { return sysCall('store.keys', { store: name || '' }); },
    },
    os: {
      info: function () { return sysCall('os.info', {}); }, // {platform,arch,hostname,osVersion,goVersion,numCPU}
    },
    process: {
      id: function () { return sysCall('process.id', {}); },
      exit: function (code) { return sysCall('process.exit', { code: code || 0 }); },
      restart: function () { return sysCall('process.restart', {}); }, // 拉起新实例后退出的安全重启
    },
    // ---- G4 自动更新（对标 Tauri updater）----
    // 检查/安装是异步长任务：调用只发起，结果经事件回推——
    //   'update.available'({version,url,sha256,notes}) / 'update.upToDate'({current})
    //   / 'update.installed'({version,restartRequired}) / 'update.error'({message})
    // install 只接受 check 时已通过 ed25519 验签的缓存条目；换装后需 process.restart 生效。
    update: {
      check: function () { return sysCall('update.check', {}); },
      install: function () { return sysCall('update.install', {}); },
      pending: function () { return sysCall('update.pending', {}); },
      onAvailable: function (cb) { return on('update.available', cb); },
      onUpToDate: function (cb) { return on('update.upToDate', cb); },
      onInstalled: function (cb) { return on('update.installed', cb); },
      onError: function (cb) { return on('update.error', cb); },
    },
    taskbar: {
      setProgress: function (value) { return sysCall('taskbar.progress', { value: value }); },
      setState: function (state) { return sysCall('taskbar.state', { state: state }); }, // normal|paused|error|indeterminate
      clear: function () { return sysCall('taskbar.clear', {}); },
      setOverlay: function (iconDataURL) { return sysCall('taskbar.overlay', { icon: iconDataURL }); },
      clearOverlay: function () { return sysCall('taskbar.clearOverlay', {}); },
    },
    dialog: {
      message: function (opts) { return sysCall('dialog.message', opts || {}); },
      open: function (opts) { return sysCall('dialog.open', opts || {}); },
      save: function (opts) { return sysCall('dialog.save', opts || {}); },
    },
    tray: {
      // create({icon: dataURL, tooltip})；事件：tray:click / tray:double-click / tray:menu({id})。
      create: function (opts) { return trayCall('tray.create', opts || {}); },
      destroy: function () { return trayCall('tray.destroy', {}); },
      setTooltip: function (text) { return trayCall('tray.tooltip', { tooltip: text }); },
      setMenu: function (items) { return trayCall('tray.menu', { items: items }); },
    },
    menu: {
      // 原生菜单栏（挂主窗口）：items 同 tray.setMenu 结构；点击走 'tray:menu' 事件。
      set: function (items) { return trayCall('menu.set', { items: items }); },
    },
  };

  function trayCall(method, args) {
    if (typeof window.__freedom_tray !== 'function') {
      return Promise.reject(new Error('[freedom] 未检测到托盘桥接。'));
    }
    return window.__freedom_tray(method, JSON.stringify(args || {}));
  }

  // sys 桥接：原生系统能力（taskbar.* / window.* / dialog.*）。
  // W5 会扩展为 freedom.sys 命名空间，先行提供内部包装。
  function sysCall(method, args) {
    if (typeof window.__freedom_sys !== 'function') {
      return Promise.reject(new Error('[freedom] 未检测到原生系统能力桥接。'));
    }
    return window.__freedom_sys(method, JSON.stringify(args || {}));
  }

  // 前端自绘按钮的便捷绑定：把按钮 DOM 接到窗口控制。
  // freedom.window.bindButtons({ min: '#minBtn', max: '#maxBtn', close: '#closeBtn' })
  freedom.window.bindButtons = function (sel) {
    var q = function (s) { return typeof s === 'string' ? document.querySelector(s) : s; };
    var min = q(sel && sel.min), max = q(sel && sel.max), close = q(sel && sel.close);
    var self = this;
    var ignoreErr = function () { /* 桥接未就绪（如窗口销毁中）时忽略本次操作 */ };
    if (min) min.addEventListener('click', function () { self.minimize().catch(ignoreErr); });
    if (max) max.addEventListener('click', function () {
      self.isMaximized().then(function (m) {
        if (m) self.unmaximize(); else self.maximize();
      }).catch(ignoreErr);
    });
    if (close) close.addEventListener('click', function () { self.close().catch(ignoreErr); });
  };

  function windowAction(action, args) {
    // __freedom_window 是 webview_go Bind 注入的异步全局函数（返回 Promise）。
    // Go 侧签名为 (action, paramsJSON)，paramsJSON 恒为 JSON object 字符串。
    if (typeof window.__freedom_window !== 'function') {
      return Promise.reject(new Error('[freedom] 未检测到原生窗口控制桥接。'));
    }
    return window.__freedom_window(action, JSON.stringify(args || {}));
  }

  Object.defineProperty(freedom, 'isDesktop', {
    get: function () { return !!getBridge(); },
  });

  // 兼容 Wails 风格：window.go 对象，便于迁移既有代码。
  var go = {};
  Object.defineProperty(go, 'backend', {
    get: function () { return { call: call, invoke: call }; },
  });
  window.go = go;

  window.freedom = freedom;

  // H3：页面就绪回调。Go 侧 onReady 改为在 __freedom__ready 被调用时触发
  //（此前在 SetHtml 前触发，期间 Emit 的初始化事件因 SDK 尚未建立而丢失）。
  // DOMContentLoaded 后上报就绪；若脚本执行时 DOM 已加载完成则立即上报。
  function signalReady() {
    if (typeof window.__freedom__ready === 'function') {
      try { window.__freedom__ready(); } catch (e) { /* 忽略 */ }
    }
  }
  if (document.readyState === 'loading') {
    document.addEventListener('DOMContentLoaded', signalReady);
  } else {
    signalReady();
  }
})();
